### PR TITLE
feat: add read and verify utility method

### LINKED ISSUE #306: Utility method for verifying content from reader
When reading from the `io.ReadCloser` returned by `Fetch()` functions, it's recommended to verify the read content against the descriptor.
`oras-go` should provide a convenient utility method for verifying the content read from reader.

### PR DESCRIPTION
**Pull request by wangxiaoxuan273**
Added VerifyReader struct that performs verification against the digest and size when reading content. Changed `CopyBuffer` to use `VerifyReader`.

resolves #306 

Signed-off-by: wangxiaoxuan273 [wangxiaoxuan119@gmail.com](mailto:wangxiaoxuan119@gmail.com)

### CODE REVIEW

File: content/reader.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+// using a VerifyReader.
+func ReadAll(r io.Reader, desc ocispec.Descriptor) ([]byte, error) {
+	if desc.Size < 0 {
+		return nil, ErrInvalidDescriptorSize
+	}
+	buf := make([]byte, desc.Size)
+
+	vr := NewVerifyReader(r, desc)
+	if _, err := io.ReadFull(vr, buf); err != nil {
+		return nil, fmt.Errorf("read failed: %w", err)
```
Comment (wangxiaoxuan273) [inline content/reader.go:118]: Should this error be made into an error variable, as `ErrInvalidDescriptorSize`?

**Comment by wangxiaoxuan273** [comment]
Should I add unit tests to `CopyBuffer`? I'm afraid I'll totally break it.

File: content/reader.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+	}
+	if err := vr.Verify(); err != nil {
+		return nil, err
+	}
+	return buf, nil
+}
+
+// EnsureEOF ensures the read operation ends with an EOF and no
+// trailing data is present.
+func EnsureEOF(r io.Reader) error {
```
Comment (Wwwsylvia) [inline content/reader.go:128]: Can we keep this private?

File: content/reader.go (hunk 1 above)
Comment (Wwwsylvia) [reply content/reader.go:118]: The descriptor size is valid, just that it does not match with the actual content size. I think we can keep it as it is.

File: content/reader.go (hunk 2 above)
Comment (shizhMSFT) [inline content/reader.go:128]: Can we make this method private?

File: content/reader.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+	var peek [1]byte
+	_, err := io.ReadFull(r, peek[:])
+	if err != io.EOF {
+		return errors.New("trailing data")
```
Comment (shizhMSFT) [inline content/reader.go:132]:
```suggestion
		return ErrTrailingData
```

File: internal/ioutil/io.go [hunk 4]
```diff
@@ -37,31 +37,14 @@ func (fn CloserFunc) Close() error {
...
-// trailing data is present.
-func EnsureEOF(r io.Reader) error {
-	var peek [1]byte
-	_, err := io.ReadFull(r, peek[:])
-	if err != io.EOF {
-		return errors.New("trailing data")
-	}
-
-	return nil
+	return content.EnsureEOF(vr)
```
Comment (Wwwsylvia) [inline internal/ioutil/io.go:47]: Do we still need this? EOF is already ensured in `Verify()`.

File: content/reader.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+	if vr.err == nil {
+		if vr.base.N > 0 {
+			return errors.New("early verify")
+		}
+	} else if vr.err != io.EOF {
+		return vr.err
+	}
+
+	if err := EnsureEOF(vr.base.R); err != nil {
+		vr.err = ErrTrailingData
```
Comment (shizhMSFT) [inline content/reader.go:81]:
```suggestion
		vr.err = err
```

File: internal/ioutil/io.go (hunk 4 above)
Comment (shizhMSFT) [inline internal/ioutil/io.go:47]: We don't need to check EOF here since it has already checked by the `VerifyReader()`.

File: content/reader.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+	ErrMismatchedDigest = errors.New("mismatched digest")
+
+	// ErrTrailingData is returned by ReadAll() when
+	// there exists trailing data unread when the read terminates.
+	ErrTrailingData = errors.New("trailing data")
+)
+
+// VerifyReader reads the content described by its descriptor and verifies
+// against its size and digest.
+type VerifyReader struct {
```
Comment (shizhMSFT) [inline content/reader.go:43]: Could you add an example for `VerifyReader`?

File: content/reader.go (hunk 1 above)
Comment (shizhMSFT) [reply content/reader.go:118]: No, we should keep the original error.

File: content/reader_test.go [hunk 7]
```diff
@@ -26,7 +26,83 @@ import (
...
+	buf := make([]byte, 5)
+	n, err := vr.Read(buf)
+
+	if !bytes.Equal(buf, []byte("examp")) {
+		t.Fatalf("incorrect read content: %s", buf)
+	}
+	if n != 5 {
+		t.Fatalf("incorrect number of bytes read: %d", n)
+	}
+	if err != nil {
```
Comment (Wwwsylvia) [inline content/reader_test.go:48]: We should always check the err first.

**Review by shizhMSFT — CHANGES_REQUESTED** [review]
We need more test cases for edge cases

File: content/reader_test.go [hunk 8]
```diff
@@ -26,7 +26,83 @@ import (
...
+
+	// mismatched content and descriptor
+	r = bytes.NewReader([]byte("foo"))
+	vr = NewVerifyReader(r, desc)
+	buf = make([]byte, 5)
+	n, err = vr.Read(buf)
+	if n != 3 {
+		t.Fatalf("incorrect number of bytes read: %d", n)
+	}
+	if err != nil {
```
Comment (Wwwsylvia) [inline content/reader_test.go:60]: Same here

File: internal/ioutil/io.go (hunk 4 above)
Comment (wangxiaoxuan273) [reply internal/ioutil/io.go:47]: removed it.

File: content/reader.go (hunk 2 above)
Comment (wangxiaoxuan273) [reply content/reader.go:128]: done

File: internal/ioutil/io.go (hunk 4 above)
Comment (wangxiaoxuan273) [reply internal/ioutil/io.go:47]: removed it

File: content/reader.go (hunk 2 above)
Comment (wangxiaoxuan273) [reply content/reader.go:128]: done.

File: content/reader.go (hunk 3 above)
Comment (wangxiaoxuan273) [reply content/reader.go:132]: Changed accordingly.

File: content/reader.go (hunk 5 above)
Comment (wangxiaoxuan273) [reply content/reader.go:81]: changed.

File: content/reader_test.go (hunk 7 above)
Comment (wangxiaoxuan273) [reply content/reader_test.go:48]: Moved it to first.

File: content/reader_test.go (hunk 8 above)
Comment (wangxiaoxuan273) [reply content/reader_test.go:60]: moved it to the first.

**Comment by wangxiaoxuan273** [comment]
> Should I add unit tests to `CopyBuffer`? I'm afraid I'll totally break it.

I will add some unit tests to `CopyBuffer`.

File: internal/ioutil/io.go [hunk 9]
```diff
@@ -37,30 +37,13 @@ func (fn CloserFunc) Close() error {
...
-
-	if _, err := io.CopyBuffer(mw, lr, buf); err != nil {
+	vr := content.NewVerifyReader(src, desc)
+	if _, err := io.CopyBuffer(dst, vr, buf); err != nil {
 		return fmt.Errorf("copy failed: %w", err)
 	}
-
-	if !verifier.Verified() {
-		return errors.New("digest verification failed")
+	if err := vr.Verify(); err != nil {
```
Comment (shizhMSFT) [inline internal/ioutil/io.go:44]: nit:
```suggestion
	return vr.Verify()
```

File: content/reader.go (hunk 6 above)
Comment (wangxiaoxuan273) [reply content/reader.go:43]: Added one example.

File: internal/ioutil/io.go (hunk 9 above)
Comment (wangxiaoxuan273) [reply internal/ioutil/io.go:44]: Changed.

File: content/example_test.go [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+	"oras.land/oras-go/v2/content"
+)
+
+// ExampleVerifyReader gives an example of creating and using VerifyReader.
+func ExampleVerifyReader() {
+	blob := []byte("hello world")
+	desc := ocispec.Descriptor{
+		MediaType: ocispec.MediaTypeImageLayer,
+		Digest:    digest.FromBytes(blob),
+		Size:      int64(len(blob))}
```
Comment (shizhMSFT) [inline content/example_test.go:33]: We can use [content.NewDescriptorFromBytes()](https://pkg.go.dev/oras.land/oras-go/v2@v2.0.0-rc.3.0.20220914090833-524c614b5825/content#NewDescriptorFromBytes).

File: content/example_test.go [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+	r := bytes.NewReader(blob)
+	vr := content.NewVerifyReader(r, desc)
+	buf := make([]byte, len(blob))
+
+	if _, err := vr.Read(buf); err != nil {
+		panic(err)
+	}
```
Comment (shizhMSFT) [inline content/example_test.go:40]: It is better to use `bytes.Buffer` with `Buffer.ReadFrom`.

File: content/example_test.go [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+	if err := vr.Verify(); err != nil {
+		panic(err)
+	}
+
+	fmt.Println(string(buf))
```
Comment (shizhMSFT) [inline content/example_test.go:45]: Please add a comment stating that users should not trust the read content only if `Verify()` returns nil.

File: internal/ioutil/io_test.go [hunk 13]
```diff
@@ -50,3 +54,62 @@ func TestUnwrapNopCloser(t *testing.T) {
...
+	}
+	tests := []struct {
+		name    string
+		args    args
+		wantDst string
+		wantErr error
+	}{
+		{
+			name:    "exact buffer size, no errors",
+			args:    args{bytes.NewReader([]byte("foo")), make([]byte, 3), content.NewDescriptorFromBytes("test", []byte("foo"))},
```
Comment (Wwwsylvia) [inline internal/ioutil/io_test.go:72]: nit: should we make `[]byte("foo")` a variable?

File: internal/ioutil/io_test.go [hunk 14]
```diff
@@ -50,3 +54,62 @@ func TestUnwrapNopCloser(t *testing.T) {
 		})
 	}
 }
+
+func TestCopyBuffer(t *testing.T) {
```
Comment (shizhMSFT) [inline internal/ioutil/io_test.go:58]: Missing a test of unexpected EOF.

File: content/reader_test.go [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+func TestVerifyReader_Verify(t *testing.T) {
+	// matched content and descriptor
+	content := []byte("example content")
+	desc := ocispec.Descriptor{
+		MediaType: ocispec.MediaTypeImageLayer,
+		Digest:    digest.FromBytes(content),
+		Size:      int64(len(content))}
+	r := bytes.NewReader(content)
+	vr := NewVerifyReader(r, desc)
+	buf := make([]byte, len(content))
```
Comment (shizhMSFT) [inline content/reader_test.go:90]: The content of `buf` should be also checked in the happy path.

File: content/reader_test.go [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+	n, err = vr.Read(buf)
+	if err != nil {
+		t.Fatal("Read() error = ", err)
+	}
+	if n != 3 {
+		t.Fatalf("incorrect number of bytes read: %d", n)
+	}
+}
+
+func TestVerifyReader_Verify(t *testing.T) {
```
Comment (shizhMSFT) [inline content/reader_test.go:81]: missing case of early verify and early EOF.

File: internal/ioutil/io_test.go [hunk 17]
```diff
@@ -50,3 +54,62 @@ func TestUnwrapNopCloser(t *testing.T) {
...
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			dst := &bytes.Buffer{}
+			err := CopyBuffer(dst, tt.args.src, tt.args.buf, tt.args.desc)
+			if err != tt.wantErr {
+				t.Errorf("CopyBuffer() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			gotDst := dst.String()
+			if err == nil && gotDst != tt.wantDst {
```
Comment (Wwwsylvia) [inline internal/ioutil/io_test.go:110]: Do we need to check `err == nil`?

File: content/example_test.go (hunk 10 above)
Comment (Wwwsylvia) [inline content/example_test.go:33]: Should we leverage `NewDescriptorFromBytes`?

File: content/reader_test.go [hunk 18]
```diff
@@ -0,0 +1,225 @@
...
+	content := []byte("example content")
+	desc := ocispec.Descriptor{
+		MediaType: ocispec.MediaTypeImageLayer,
+		Digest:    digest.FromBytes(content),
+		Size:      int64(len(content))}
+	r := bytes.NewReader(content)
+	vr := NewVerifyReader(r, desc)
+	buf := make([]byte, 5)
+	n, err := vr.Read(buf)
+
```
Comment (Wwwsylvia) [inline content/reader_test.go:37]: nit: empty line

File: content/reader_test.go [hunk 19]
```diff
@@ -0,0 +1,225 @@
...
+		Digest:    digest.FromBytes(content),
+		Size:      int64(len(content))}
+	r = bytes.NewReader(content)
+	vr = NewVerifyReader(r, desc)
+	buf = make([]byte, len(content))
+	n, err = vr.Read(buf)
+	if err != nil {
+		t.Fatal("Read() error = ", err)
+	}
+	if n != len(content) {
```
Comment (Wwwsylvia) [inline content/reader_test.go:58]: Do we want to compare the content also?

File: content/example_test.go (hunk 10 above)
Comment (wangxiaoxuan273) [reply content/example_test.go:33]: changed accordingly.

File: content/example_test.go (hunk 12 above)
Comment (wangxiaoxuan273) [reply content/example_test.go:45]: Added the comment.

File: content/example_test.go (hunk 10 above)
Comment (wangxiaoxuan273) [reply content/example_test.go:33]: Good point. Changed accordingly.

File: content/reader_test.go (hunk 18 above)
Comment (wangxiaoxuan273) [reply content/reader_test.go:37]: deleted.

File: internal/ioutil/io_test.go (hunk 17 above)
Comment (wangxiaoxuan273) [reply internal/ioutil/io_test.go:110]: After discussion, we think we need this check.

File: internal/ioutil/io_test.go (hunk 13 above)
Comment (wangxiaoxuan273) [reply internal/ioutil/io_test.go:72]: Added this variable.

File: content/example_test.go (hunk 11 above)
Comment (wangxiaoxuan273) [reply content/example_test.go:36]: Changed according to our offline discussion.

File: internal/ioutil/io_test.go (hunk 14 above)
Comment (wangxiaoxuan273) [reply internal/ioutil/io_test.go:58]: Added a test of unexpected EOF.

File: content/reader_test.go (hunk 16 above)
Comment (wangxiaoxuan273) [reply content/reader_test.go:74]: Added the case for early verify.

File: content/reader_test.go (hunk 15 above)
Comment (wangxiaoxuan273) [reply content/reader_test.go:80]: Added the check for `buf` content.

File: content/reader_test.go (hunk 19 above)
Comment (wangxiaoxuan273) [reply content/reader_test.go:57]: Yes I think we do. Added content check here.

**Review by shizhMSFT — APPROVED** [review]
LGTM
